lib/SILOptimizer/Transforms/SILCodeMotion.cpp

```diff
@@ -292,6 +293,44 @@ void BBEnumTagDataflowState::handlePredSwitchEnum(SwitchEnumInst *S) {
                    "the switch_enum.");
 }
 
+void BBEnumTagDataflowState::handleStringCmp(SILBasicBlock *Pred) {
+  // Find all apply instructions
+  for (auto begin = Pred->begin(); begin != Pred->end(); ++begin) {
+    if (auto *AI = dyn_cast<ApplyInst>(begin)) {
+      if (auto *FN = dyn_cast<FunctionRefInst>(AI->getCalleeOrigin())) {
+        // Only keep going if this is an apply instruction
+        if (FN->getReferencedFunctionOrNull()->getName() != "$sSS2eeoiySbSS_SStFZ") continue;
```

Comment:
Our usual way of flagging functions for SIL optimization is to use the `@_semantics` attribute to put a unique string tag on the declaration, and to have the optimizer look for that tag instead of for a specific symbol name. `@_semantics` will delay inlining the function so that higher-level optimizations have a change to apply first, and makes the pass less sensitive to changes in the standard library itself that might affect mangled names.

Reply:
Do you know what function I should apply the `@_semantics` decorator to? Is [`_stringCompare`](https://github.com/apple/swift/blob/e9d4687e31a3ae8e90604d3b15bf8b241479c211/stdlib/public/core/StringComparison.swift#L17) always the function used to compare strings?

Reply:
@milseman would know best.

Reply:
` _stringCompare` works, though it's already inlinable for the same-struct-bits fast-path.

Reply:
Will this optimization sometimes be missed if `_stringCompare` is inlined, then?

Reply:
@zoecarver What you say is possible. On the upside, inlining may also result in the code getting optimized through other optimizations. But, on a different note, it is probably more important to *not* to compare the mangled name of `_stringCompare` in the code, but add an @_semantics attribute to that function and use that instead (as suggested by others earlier). 

But, one thing to note is that, adding a @_semantics attribute to a function will interfere with the inlining of the function, which could make uses of this function in other contexts slower. So, I would think it is best to create a new PR that just adds  @_semantics attribute, and run benchmark test on it to see if there are any performance regressions. Once that lands, you can use it to modify this code here.

Reply:
It looks like the string compares I had been optimizing were not actually using `_stringCompare` and instead used another function that took the string literals directly as arguments. `_stringCompare` takes `_StringGuts` types which require store/load. That makes it much more difficult to extract the string literal. Any ideas as to the function it is calling instead? Is it OK to rely on it always optimizing into this function?

Reply:
Turns out the function it calls is just the `op==`. Looking for `hasSemanticsAttr("string.equals")` works for both if and switch statements. Using that fixes the concern about inlining `_stringCompare` too. Any reasons not to just look for `string.equals`?

Reply:
If the semantics tag already exists, it seems reasonable to use it here.